food-delivery/packages/workflows/order.ts

```diff
@@ -0,0 +1,98 @@
+import {
+  defineSignal,
+  defineQuery,
+  setHandler,
+  condition,
+  ApplicationFailure,
+  sleep,
+  proxyActivities,
+} from '@temporalio/workflow'
+import { errorMessage, getProductById, Product } from 'common'
+import type * as activities from 'activities'
+
+type OrderState = 'Charging card' | 'Paid' | 'Picked up' | 'Delivered'
+
+export interface OrderStatus {
+  productId: number
+  state: OrderState
+  deliveredAt?: Date
+}
+
+export const pickedUp = defineSignal('pickedUp')
+export const delivered = defineSignal('delivered')
+export const getStatus = defineQuery<OrderStatus>('getStatus')
+
+const { chargeCustomer, refundOrder, sendPushNotification } = proxyActivities<typeof activities>({
+  startToCloseTimeout: '1m',
+  retry: {
+    maximumInterval: '5s',
+  },
+})
+
+export async function order(productId: number): Promise<void> {
+  // validate input
+  const product = getProductById(productId)
+  if (!product) {
+    throw ApplicationFailure.create({ message: `Product ${productId} not found` })
+  }
+
+  // declare variables
+  let state: OrderState = 'Charging card'
+  let deliveredAt: Date
+
+  // setup Signal and Query handlers
+  setHandler(pickedUp, () => {
+    if (state === 'Paid') {
+      state = 'Picked up'
+    }
+  })
+
+  setHandler(delivered, () => {
+    if (state === 'Picked up') {
+      state = 'Delivered'
+      deliveredAt = new Date()
+    }
+  })
+
+  setHandler(getStatus, () => {
+    return { state, deliveredAt, productId }
+  })
+
+  // business logic
+  try {
+    await chargeCustomer(product)
+  } catch (err) {
+    const message = `Failed to charge customer for ${product.name}. Error: ${errorMessage(err)}`
+    await sendPushNotification(message)
+    throw ApplicationFailure.create({ message })
+  }
+
+  state = 'Paid'
+
+  const notPickedUpInTime = !(await condition(() => state === 'Picked up', '1 min'))
+  if (notPickedUpInTime) {
+    await refundAndNotify(
+      product,
+      '⚠️ No drivers were available to pick up your order. Your payment has been refunded.'
+    )
```

Comment:
You should transition to a `Refunding` state here to ending up in `Picked up` state in case a driver decides to pick this order up between the time this signal was received and the notification is sent.
How does this app deal with edge cases like this? There doesn't seem to be a way to notify the driver of order state changes.

Reply:
Adding Refunding, thanks. Currently no driver notifications, will be adding to this sample for later blog posts. First one is a short 101.